### PR TITLE
New Pytools package: pyfortran (with utility to check subroutine usage in a fortran module)

### LINKED ISSUE #921: meta: detect unused routines in use statements
Utility to find unused routines in used statements
https://github.com/hkershaw-brown/futils/blob/main/check_fortran_module_usage.py

```
 python check_fortran_module_usage.py ../DART/models/ROMS/model_mod.f90
Routines from '../DART/models/ROMS/model_mod.f90' NOT USED:
DEG2RAD
E_WARN
MISSING_R4
PI
QTY_POTENTIAL_TEMPERATURE
QTY_SEA_SURFACE_PRESSURE
SECPERDAY
file_to_text
find_textfile_dims
get_location
get_state
get_unit
nc_get_location_varids
nc_write_location
nc_write_location_atts
r4
rad2deg
write_location
```

### PR DESCRIPTION
**Pull request by mjs2369**
## Description:

Adds a new python package to DART pytools - pyfortran

This package is available on the Python Package Index and can be installed with the rest of pytools, or alternatively completely independent of DART with pip.

pyfortran includes one fortran-based utility - a program called check_fortran_module_usage that checks for unused subroutines in a given Fortran module.'

pyfortran can have additional utilities/programs added to it as they are developed.

A test for the functions in check_fortran_module_usage has been included as well.

### Fixes issue
fixes #921 

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

### Documentation changes needed?

- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly.

### Tests
Please describe any tests you ran to verify your changes.

- Ran the check_fortan_module_usage program with various DART fortran modules and and verified the results were correct.
- Ran the test pytools/pyfortran/tests/test_fortran_module_usage.py and ensured all tests were passing.
- Installed the pytools package suite with the instructions found at [https://github.com/NCAR/DART/blob/a10b81e8dbf428250b3cbe1f3f078a0d8f1b0465/guide/pytools.rst](https://github.com/NCAR/DART/blob/a10b81e8dbf428250b3cbe1f3f078a0d8f1b0465/guide/pytools.rst) and ran the check_fortan_module_usage executable. Ensured using the --help argument was working.
- Installed pyfortran as an individual package with pip and verified the executable was working properly.

## Checklist for merging

- [ ] Updated changelog entry
- [ ] Documentation updated
- [ ] Update conf.py

## Checklist for release
- [ ] Merge into main
- [ ] Create release from the main branch with appropriate tag
- [ ] Delete feature-branch

## Testing Datasets

- [ ] Dataset needed for testing available upon request
- [ ] Dataset download instructions included
- [x] No dataset needed

### CODE REVIEW

File: pytools/pyfortran/tests/test_fortran_module_usage.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        "subroutine test1",
+        "end subroutine test1",
+        "subroutine test2",
+        "call test1",
+        "end subroutine test2",
+    ]
+    unused = find_unused_subroutines(lines)
+    assert unused == ["test2"]
+
+def test_find_unused_routines_from_other_modules():
```
Comment (hkershaw-brown) [inline pytools/pyfortran/tests/test_fortran_module_usage.py:30]: Does check_fortran_module_usage support 
```
use routine1 & ! comment about routine
      routine2
```

If so, tt would be good to have a test case for this

Comment (mjs2369) [reply pytools/pyfortran/tests/test_fortran_module_usage.py:30]: @hkershaw-brown It does support this, but copilot left a review comment about how it thinks I shouldn't be allowing for that.

"It should only treat lines ending with '&' as continuation lines, as '&' can appear in other contexts (like string literals or comments)"

What do you think?

File: pytools/pyfortran/src/fortran_module_usage/check_fortran_module_usage.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    subroutines = []
+    for idx, line in enumerate(lines):
+        m = subroutine_pattern.match(line)
+        if m:
+            subroutines.append(m.group(1))
+
+    # Check if each subroutine is called in the file
+    usage = defaultdict(bool)
+    for subroutine in subroutines:
+        pattern = re.compile(r'\bcall {}\b'.format(re.escape(subroutine)))
```
[the remark this answers is not in the document]
Comment (mjs2369) [reply pytools/pyfortran/src/fortran_module_usage/check_fortran_module_usage.py:44]: My pattern already contains this whitespace as-is. 

Output from print statements added to prove this:
```
pattern:  re.compile('\\bcall write_time\\b')
found call line:  if (present(unit))     call write_time(unit, label, string1, string2, string3, tz, brief
```

File: pytools/pyfortran/tests/test_fortran_module_usage.py (hunk 1 above)
Comment (hkershaw-brown) [reply pytools/pyfortran/tests/test_fortran_module_usage.py:30]: I think we should take the copilot reviews with a grain of salt.  My bad, I've been playing with copilot reviews to see how well it does 

Your new test is along the lines of what I was thinking about, people commenting in use statements. 
We're not writing a parser for a compiler, so I'm not overly worried if this tool occasionally hits edge cases where people have been over enthusiastic with ampersands in their comments.
